oap-server/server-receiver-plugin/otel-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/otel/otlp/OpenTelemetryMetricRequestProcessor.java

```diff
@@ -161,6 +163,54 @@ private static Map<Double, Long> buildBuckets(
         return result;
     }
 
+    /**
+     * ExponentialHistogram data points are an alternate representation to the Histogram data point in OpenTelemetry
+     * metric format(https://opentelemetry.io/docs/reference/specification/metrics/data-model/#exponentialhistogram).
+     * It uses scale, offset and bucket index to calculate the bound. Firstly, calculate the base using scale by
+     * formula: base = 2**(2**(-scale)). Then the upperBound of specific bucket can be calculated by formula:
+     * base**(offset+index+1). Above calculation way is about positive buckets. For the negative case, we just
+     * map them by their absolute value into the negative range using the same scale as the positive range. So the
+     * upperBound should be calculated as -base**(offset+index).
+     *
+     * @param positiveOffset       corresponding to positive Buckets' offset in ExponentialHistogramDataPoint
+     * @param positiveBucketCounts corresponding to positive Buckets' bucket_counts in ExponentialHistogramDataPoint
+     * @param negativeOffset       corresponding to negative Buckets' offset in ExponentialHistogramDataPoint
+     * @param negativeBucketCounts corresponding to negative Buckets' bucket_counts in ExponentialHistogramDataPoint
+     * @param scale                corresponding to scale in ExponentialHistogramDataPoint
+     * @return The map is a bucket set for histogram, the key is specific bucket's upperBound, the value is item count
+     * in this bucket lower than or equals to key(upperBound)
+     */
+    private static Map<Double, Long> buildBucketsFromExponentialHistogram(
+        int positiveOffset, final List<Long> positiveBucketCounts,
+        int negativeOffset, final List<Long> negativeBucketCounts, int scale) {
+
+        final Map<Double, Long> result = new HashMap<>();
```

Comment:
Looks like you didn't take into consideration of the `zero_count` field in OTEL exponential histgram data model? Can you elaborate why?

Reply:
I think I miss this part, I would like to assign its upper bound as the lower bound of small positive, WDYT?

Reply:
From my understanding, zero means less than a threshold, no matter this is less than which boundaries.

Reply:
> I think I miss this part, I would like to assign its upper bound as the lower bound of small positive, WDYT?

> From my understanding, zero means less than a threshold, no matter this is less than which boundaries.

There is clear definition of zeros in OTEL data model, we can tackle the zeros according to the definition, in short, the `zero_count` is the count of bucket `(-threshold, +threshold)`, or count of samples that are approximately `0` if `threshold` is not set.

> The ExponentialHistogram contains a special zero_count bucket and an optional zero_threshold field where zero_count contains the count of values whose absolute value is less than or equal to zero_threshold. The precise value for the zero_threshold is arbitrary and not related to the scale.

> When zero_threshold is unset or 0, this bucket stores values that cannot be expressed using the standard exponential formula as well as values that have been rounded to zero.

Reply:
Ignoring the count make sense when do percentile, as it should not be considered as base.
But if we read the count, maybe this helps.
This is just my understanding.

Reply:
> There is clear definition of zeros in OTEL data model, we can tackle the zeros according to the definition, in short, the zero_count is the count of bucket (-threshold, +threshold), or count of samples that are approximately 0 if threshold is not set.

Yes, they defined. Last time, I check, it mentioned, the zero threshold even could cover over the existing bucket scopes.

Reply:
Check this https://opentelemetry.io/docs/reference/specification/metrics/data-model/#exponentialhistogram

> The ExponentialHistogram contains a special zero_count bucket and an optional zero_threshold field where zero_count contains the count of values whose absolute value is less than or equal to zero_threshold. The precise value for the zero_threshold is arbitrary and not related to the scale.

So, we can't assume the range of the threshold would not overlap with existing buckets. We can't build more buckets from it.

Reply:
We could leave some comments here as a TBD about ignoring this, maybe one day, we will know how should we process this. WDYT?

Reply:
> We could leave some comments here as a TBD about ignoring this, maybe one day, we will know how should we process this. WDYT?

Good to me

Reply:
> We could leave some comments here as a TBD about ignoring this, maybe one day, we will know how should we process this. WDYT?

I'm back from class. I read the docs and the new opentelemetry proto file. I think it is a nice idea.

Reply:
About comments, should I add them to code?

Reply:
> Check this https://opentelemetry.io/docs/reference/specification/metrics/data-model/#exponentialhistogram
> 
> 
> 
> > The ExponentialHistogram contains a special zero_count bucket and an optional zero_threshold field where zero_count contains the count of values whose absolute value is less than or equal to zero_threshold. The precise value for the zero_threshold is arbitrary and not related to the scale.
> 
> 
> 
> So, we can't assume the range of the threshold would not overlap with existing buckets. We can't build more buckets from it.

Actually I am curious where does it indicate the threshold can overlap with existing buckets ?

Reply:
> The precise value for the zero_threshold is arbitrary and not related to the scale.

**not related to the scale** means **not related to the bucket**. The buckets are made by (2**2**-scale)